rskj-core/src/main/java/co/rsk/core/bc/MiningMainchainViewImpl.java

```diff
@@ -73,24 +79,84 @@ public void addBest(BlockHeader bestHeader) {
-            newHeaderChain.add(currentHeader);
+    /**
+     * Given a new best header and the index of its parent rebuild the mainchain using the index as the pivot point
+     * by discarding all headers subsequent to the it, setting the new header as the tip and refilling with as many
+     * are needed to complete the desired height
+     *
+     * @param bestHeader The best header to be on top of the chain
+     * @param parentIndex List index of the best header's parent
+     */
+    private void addBestAndRebuildFromParent(BlockHeader bestHeader, int parentIndex) {
+        List<BlockHeader> staleHeaders = mainchain.stream()
+                .limit(parentIndex)
+                .collect(Collectors.toList());
+
+        List<BlockHeader> commonAncestorChain = mainchain.stream()
+                .skip(staleHeaders.size())
+                .collect(Collectors.toList());
+
+        List<BlockHeader> newMainchain = Stream.concat(
+                Arrays.asList(bestHeader).stream(),
+                commonAncestorChain.stream())
+                .collect(Collectors.toList());
+
+        mainchain = newMainchain;
+
+        fillMissingHeaders();
+        deleteHeaders(staleHeaders);
+    }
+
+    /**
+     * Fill the mainchain's tail with as many headers as it's needed to satisfy the desired height
+     */
+    private void fillMissingHeaders() {
+        int currentSize = mainchain.size();
+
+        if (currentSize == 0 || currentSize >= height) {
+            return;
+        }
+
+        BlockHeader lastHeader = mainchain.get(currentSize - 1);
+
+        List<BlockHeader> missingHeaders = retrieveAncestorsForHeader(lastHeader, height - currentSize);
 
-            if(!blocksByHash.containsKey(currentHeader.getHash())) {
-                addHeaderToMaps(currentHeader);
+        for (BlockHeader header : missingHeaders) {
+            if(!blocksByHash.containsKey(header.getHash())) {
+                addHeaderToMaps(header);
             }
+        }
+
+        mainchain = Stream.concat(mainchain.stream(), missingHeaders.stream())
+                .collect(Collectors.toList());
+    }
 
+    /**
+     * Given a start block header and a chain length, retrieve a List of said lenght consisting of
+     * the start header's ancestors
+     *
+     * The returned list DOES NOT include the start header
+     *
+     * @param header The block header to look the ancestors for
+     * @param chainLength The max length of the returned ancestor chain
+     */
+    private List<BlockHeader> retrieveAncestorsForHeader(BlockHeader header, int chainLength) {
+        List<BlockHeader> missingHeaders = new ArrayList<>(chainLength);
+
+        BlockHeader currentHeader = header;
+        for(int i = 0; i < chainLength; i++) {
+
+            // genesis has no parent
             if(currentHeader.isGenesis()) {
                 break;
             }
 
             currentHeader = blockStore.getBlockByHash(currentHeader.getParentHash().getBytes()).getHeader();
```

Comment:
Remember to check if currentHeader is != null

Reply:
Should be covered by 50f4ceb